sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs

```diff
@@ -57,42 +58,97 @@ protected override void Initialize(SonarAnalysisContext context)
 
                     var literal = (LiteralExpressionSyntax)c.Node;
 
-                    var numberWithoutSuffix = ClearNumberTypeSuffix(literal.Token.Text);
-
-                    var decimalParts = numberWithoutSuffix.Split('.');
-                    if (decimalParts.Length > 2)
-                    {
-                        return;
-                    }
-
-                    var hasIrregularPattern = decimalParts
-                        .SelectMany(part => part.Split('_'))
-                        .Select(x => x.Length)
-                        .Skip(1) // skip the first part (1_234 => 234)
-                        .Reverse().Skip(decimalParts.Length == 2 ? 1 : 0) // skip the last if there is a decimal (.234_5 => 234)
-                        .Distinct().Skip(1).Any(); // we expect to have only 1 size of pattern
-
-                    if (hasIrregularPattern)
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        /// <remarks>internal for test purposes.</remarks>
+        internal static bool HasIrregularPattern(string numericToken)
+        {
+            var split = StripNumericPreAndSuffix(numericToken).Split(Dot);
+
+            // ignore multiple dots.
+            if (split.Length > 2)
+            {
+                return false;
+            }
+
+            var groups = split[0].Split(Underscore).Select(s => s.Length).ToArray();
```

Comment:
I suggest to rename to `groupLengths`
- this isn't "groups of digits", but the size of each groups
- it doesn't refer to all groups, but only to the ones on the left of the 'dot' (if the 'dot' is present) - *please add a comment above to mention something like* `// we are interested only in what is on the left side of the 'Dot', if present`